sgx-tools/dcap_linux.go

```diff
@@ -0,0 +1,514 @@
+}
+
+func getQuoteSizeExRequest(attKeyId []byte) (uint32, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return 0, err
+	}
+	defer conn.Close()
+
+	if len(attKeyId) != attestationKeyIdentityLength {
+		return 0, fmt.Errorf("len(attKeyId) is not %d but %d", attestationKeyIdentityLength, len(attKeyId))
+	}
+
+	req := pb.AesmServiceRequest{}
+	req.GetQuoteSizeExReq = &pb.AesmServiceRequest_GetQuoteSizeExRequest{
+		AttKeyId: attKeyId,
+		Timeout:  timeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return 0, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetQuoteSizeExRes = &pb.AesmServiceResponse_GetQuoteSizeExResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return 0, err
+	}
+
+	if resp.GetQuoteSizeExRes.GetErrorCode() != 0 {
+		return 0, fmt.Errorf("failed to get Quote Size (error code = %d)",
+			resp.GetQuoteSizeExRes.GetErrorCode())
+	}
+
+	if resp.GetQuoteSizeExRes.GetQuoteSize() == 0 {
+		return 0, fmt.Errorf("invalid Quote Size %d ",
+			resp.GetQuoteSizeExRes.GetQuoteSize())
+	}
+
+	return resp.GetQuoteSizeExRes.GetQuoteSize(), nil
+}
+
+func getQuoteExRequest(attKeyId []byte, report []byte, quoteSize uint32) ([]byte, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	if len(attKeyId) != attestationKeyIdentityLength {
+		return nil, fmt.Errorf("len(attKeyId) is not %d but %d", attestationKeyIdentityLength, len(attKeyId))
+	}
+
+	err = DumpLocalReport(report)
+	if err != nil {
+		return nil, err
+	}
+
+	req := pb.AesmServiceRequest{}
+	req.GetQuoteExReq = &pb.AesmServiceRequest_GetQuoteExRequest{
+		Report:       report,
+		AttKeyId:     attKeyId,
+		QeReportInfo: nil,
+		BufSize:      quoteSize,
+		Timeout:      timeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetQuoteExRes = &pb.AesmServiceResponse_GetQuoteExResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return nil, err
+	}
+
+	if resp.GetQuoteExRes.GetErrorCode() != 0 {
+		return nil, fmt.Errorf("failed to get Quote Size (error code = %d)",
+			resp.GetQuoteSizeExRes.GetErrorCode())
+	}
+
+	if resp.GetQuoteExRes.GetQuote() == nil {
+		return nil, fmt.Errorf("failed to get Quote")
+	}
+
+	if len(resp.GetQuoteExRes.GetQuote()) != (int)(quoteSize) {
+		return nil, fmt.Errorf("len(resp.GetQuoteExRes.GetQuote()) is not %d but %d\n", quoteSize, len(resp.GetQuoteExRes.GetQuote()))
+	}
+
+	if resp.GetQuoteExRes.GetQeReportInfo() != nil {
+		return nil, fmt.Errorf("invalid Qe Rereport Info:%v\n", resp.GetQuoteExRes.GetQeReportInfo())
+	}
+
+	return resp.GetQuoteExRes.GetQuote(), nil
+}
+
+func selectAttKeyIDRequest() ([]byte, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.SelectAttKeyIDReq = &pb.AesmServiceRequest_SelectAttKeyIDRequest{
```

Comment:
Plz leave a comment to explain why att_key_id_list is not filled.

Reply:
This problem is not resolved.

Reply:
> This problem is not resolved.

Solved already, See in the next line.